File: Arduino/epd2in7b_V2/epd2in7b_V2.cpp
/**
 *  @filename   :   epd2in7b_V2.cpp
 *  @brief      :   Implements for Dual-color e-paper library
 *  @author     :   
 *
 *  Copyright (C) Waveshare     Feb 19 2021
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documnetation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to  whom the Software is
 * furished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS OR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

#include <stdlib.h>
#include "epd2in7b_V2.h"

Epd::~Epd() {
};

Epd::Epd() {
    reset_pin = RST_PIN;
    dc_pin = DC_PIN;
    cs_pin = CS_PIN;
    busy_pin = BUSY_PIN;
    width = EPD_WIDTH;
    height = EPD_HEIGHT;
};

int Epd::Init(void) {
    /* this calls the peripheral hardware interface, see epdif */
    if (IfInit() != 0) {
        return -1;
    }
    /* EPD hardware init start */
    Reset();
     
    WaitUntilIdle();
	SendCommand(0x12);
	WaitUntilIdle();
	
    SendCommand(0x00);
    SendData(0x27);
    SendData(0x01);
	SendData(0x00);
	
    SendCommand(0x11);
    SendData(0x03); 
    
	SetWindows(0, 0, width-1, height-1);
	SetCursor(0, 0);
    /* EPD hardware init end */
    return 0;
}

/**
 *  @brief: basic function for sending commands
 */
void Epd::SendCommand(unsigned char command) {
    DigitalWrite(dc_pin, LOW);
    SpiTransfer(command);
}

/**
 *  @brief: basic function for sending data
 */
void Epd::SendData(unsigned char data) {
    DigitalWrite(dc_pin, HIGH);
    SpiTransfer(data);
}

/**
 *  @brief: Wait until the busy_pin goes HIGH
 */
void Epd::WaitUntilIdle(void) {
    while(DigitalRead(busy_pin) == 1) {      //0: busy, 1: idle
        DelayMs(10);
    }      
}

/******************************************************************************
function :	Setting the display window
parameter:
******************************************************************************/
void Epd::SetWindows(UWORD Xstart, UWORD Ystart, UWORD Xend, UWORD Yend)
{
    SendCommand(0x44); // SET_RAM_X_ADDRESS_START_END_POSITION
    SendData((Xstart>>3) & 0xFF);
    SendData((Xend>>3) & 0xFF);
	
    SendCommand(0x45); // SET_RAM_Y_ADDRESS_START_END_POSITION
    SendData(Ystart & 0xFF);
    SendData((Ystart >> 8) & 0xFF);
    SendData(Yend & 0xFF);
    SendData((Yend >> 8) & 0xFF);
}

/******************************************************************************
function :	Set Cursor
parameter:
******************************************************************************/
void Epd::SetCursor(UWORD Xstart, UWORD Ystart)
{
    SendCommand(0x4E); // SET_RAM_X_ADDRESS_COUNTER
    SendData(Xstart & 0xFF);

    SendCommand(0x4F); // SET_RAM_Y_ADDRESS_COUNTER
    SendData(Ystart & 0xFF);
    SendData((Ystart >> 8) & 0xFF);
}

/**
 *  @brief: module reset. 
 *          often used to awaken the module in deep sleep, 
 *          see Epd::Sleep();
 */
void Epd::Reset(void) {
    DigitalWrite(reset_pin, HIGH);
    DelayMs(200);   
    DigitalWrite(reset_pin, LOW);
    DelayMs(2);
    DigitalWrite(reset_pin, HIGH);
    DelayMs(200);   
}

/**
 *  @brief: transmit partial data to the black part of SRAM
 */
void Epd::TransmitPartialBlack(const unsigned char* buffer_black, int x, int y, int w, int l) {
    if (buffer_black != NULL) {
		
		int width = (w%8 == 0 ? w/8 : w/8+1);
		
		SetWindows(x, y, w+x-1, l+y-1);
		SetCursor(x, y);
        SendCommand(0x24);
        DelayMs(2);
        for(int i = 0; i < width * l; i++) {
            SendData(~buffer_black[i]);  
        }  
        DelayMs(2);
    }
}

/**
 *  @brief: transmit partial data to the red part of SRAM
 */
void Epd::TransmitPartialRed(const unsigned char* buffer_red, int x, int y, int w, int l) {
    if (buffer_red != NULL) {
		
		int width = (w%8 == 0 ? w/8 : w/8+1);
		
		SetWindows(x, y, w+x-1, l+y-1);
		SetCursor(x, y);
        SendCommand(0x26);
        DelayMs(2);
        for(int i = 0; i < width * l; i++) {
            SendData(buffer_red[i]);  
        }  
        DelayMs(2);
    }
}

/**
 * @brief: refresh and displays the frame
 */
void Epd::DisplayFrame(const unsigned char* frame_buffer_black, const unsigned char* frame_buffer_red) {
    if (frame_buffer_black != NULL) {
        SendCommand(0x24);           
        DelayMs(2);
        for(int i = 0; i < width * height / 8; i++) {
            SendData(~pgm_read_byte(&frame_buffer_black[i]));  
        }  
        DelayMs(2);
    }
    if (frame_buffer_red != NULL) {
        SendCommand(0x26);
        DelayMs(2);
        for(int i = 0; i < width * height / 8; i++) {
            SendData(pgm_read_byte(&frame_buffer_red[i]));  
        }  
        DelayMs(2);
    }
	DisplayFrame();
}

/**
 * @brief: clear the frame data from the SRAM, this won't refresh the display
 */
void Epd::ClearFrame(void) {
    SendCommand(0x24);           
    DelayMs(2);
    for(int i = 0; i < width * height / 8; i++) {
        SendData(0xff);  
    }  
    DelayMs(2);
    SendCommand(0x26);
    DelayMs(2);
    for(int i = 0; i < width * height / 8; i++) {
        SendData(0x00);  
    }  
    DelayMs(2);
	DisplayFrame();
}

/**
 * @brief: This displays the frame data from SRAM
 */
void Epd::DisplayFrame(void) {
    SendCommand(0x20); 
    WaitUntilIdle();
}

/**
 * @brief: After this command is transmitted, the chip would enter the deep-sleep mode to save power. 
 *         The deep sleep mode would return to standby by hardware reset. The only one parameter is a 
 *         check code, the command would be executed if check code = 0xA5. 
 *         You can use Epd::Reset() to awaken and use Epd::Init() to initialize.
 */
void Epd::Sleep() {
  SendCommand(0x10);
  SendData(0x01);
}

/* END OF FILE */

File: E-paper_Separate_Program/10.85inch_e-Paper_G/RaspberryPi/python/lib/waveshare_epd/epd2in7b_V2.py
# *****************************************************************************
# * | File        :	  epd2in7b_V2.py
# * | Author      :   Waveshare team
# * | Function    :   Electronic paper driver
# * | Info        :
# *----------------
# * | This version:   V1.1
# * | Date        :   2022-08-10
# # | Info        :   python demo
# -----------------------------------------------------------------------------
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documnetation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to  whom the Software is
# furished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS OR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
# THE SOFTWARE.
#


import logging
from . import epdconfig

# Display resolution
EPD_WIDTH       = 176
EPD_HEIGHT      = 264

logger = logging.getLogger(__name__)

class EPD:
    def __init__(self):
        self.reset_pin = epdconfig.RST_PIN
        self.dc_pin = epdconfig.DC_PIN
        self.busy_pin = epdconfig.BUSY_PIN
        self.cs_pin = epdconfig.CS_PIN
        self.width = EPD_WIDTH
        self.height = EPD_HEIGHT

    # Hardware reset
    def reset(self):
        epdconfig.digital_write(self.reset_pin, 1)
        epdconfig.delay_ms(200) 
        epdconfig.digital_write(self.reset_pin, 0)
        epdconfig.delay_ms(2)
        epdconfig.digital_write(self.reset_pin, 1)
        epdconfig.delay_ms(200)   

    # Send Command
    def send_command(self, command):
        epdconfig.digital_write(self.dc_pin, 0)
        epdconfig.digital_write(self.cs_pin, 0)
        epdconfig.spi_writebyte([command])
        epdconfig.digital_write(self.cs_pin, 1)

    # Send Data
    def send_data(self, data):
        epdconfig.digital_write(self.dc_pin, 1)
        epdconfig.digital_write(self.cs_pin, 0)
        epdconfig.spi_writebyte([data])
        epdconfig.digital_write(self.cs_pin, 1)

    # send a lot of data   
    def send_data2(self, data):
        epdconfig.digital_write(self.dc_pin, 1)
        epdconfig.digital_write(self.cs_pin, 0)
        epdconfig.spi_writebyte2(data)
        epdconfig.digital_write(self.cs_pin, 1)
        
    # Read Busy
    def ReadBusy(self):
        logger.debug("e-Paper busy")
        while(epdconfig.digital_read(self.busy_pin) == 1):      # 0: idle, 1: busy
            epdconfig.delay_ms(10)
        logger.debug("e-Paper busy release")
            
    # Setting the display window
    def SetWindows(self, Xstart, Ystart, Xend, Yend):
        self.send_command(0x44)
        self.send_data((Xstart >> 3) & 0xff)
        self.send_data((Xend >> 3) & 0xff)
        
        self.send_command(0x45)
        self.send_data(Ystart & 0xff)
        self.send_data((Ystart >> 8) & 0xff)
        self.send_data(Yend & 0xff)
        self.send_data((Yend >> 8) & 0xff)
    
    # Set Cursor
    def SetCursor(self, Xstart, Ystart):
        self.send_command(0x4E)
        self.send_data(Xstart & 0xff)
        self.send_command(0x4F)
        self.send_data(Ystart & 0xff)
        self.send_data((Ystart >> 8) & 0xff)
        
    # Initialize the e-Paper register
    def init(self):
        if (epdconfig.module_init() != 0):
            return -1
            
        self.reset()

        self.ReadBusy() 
        self.send_command(0x12)      
        self.ReadBusy() 
        
        self.send_command(0x00)     
        self.send_data(0x27) 
        self.send_data(0x01) 
        self.send_data(0x00) 
        
        self.send_command(0x11)     
        self.send_data(0x03) 
        
        self.SetWindows(0, 0, self.width-1, self.height-1)
        self.SetCursor(0, 0)
        return 0

    def getbuffer(self, image):
        # logger.debug("bufsiz = ",int(self.width/8) * self.height)
        buf = [0xFF] * (int(self.width/8) * self.height)
        image_monocolor = image.convert('1')
        imwidth, imheight = image_monocolor.size
        pixels = image_monocolor.load()
        # logger.debug("imwidth = %d, imheight = %d",imwidth,imheight)
        if(imwidth == self.width and imheight == self.height):
            logger.debug("Vertical")
            for y in range(imheight):
                for x in range(imwidth):
                    # Set the bits for the column of pixels at the current position.
                    if pixels[x, y] == 0:
                        buf[int((x + y * self.width) / 8)] &= ~(0x80 >> (x % 8))
        elif(imwidth == self.height and imheight == self.width):
            logger.debug("Horizontal")
            for y in range(imheight):
                for x in range(imwidth):
                    newx = y
                    newy = self.height - x - 1
                    if pixels[x, y] == 0:
                        buf[int((newx + newy*self.width) / 8)] &= ~(0x80 >> (y % 8))
        return buf
    
    # Sends the image buffer in RAM to e-Paper and displays
    def display(self, imageblack, imagered):
        Width = self.width / 8 
        Height = self.height 

        buf = [0x00] * int(Width * Height)
        for i in range(0, int(Width * Height)):
            buf[i] = ~imagered[i]

        self.send_command(0x24) 
        self.send_data2(imageblack) 

        self.send_command(0x26) 
        self.send_data2(buf) 
        
        self.TurnOnDisplay()

    # Clear the screen
    def Clear(self):
        self.send_command(0x24)
        self.send_data2([0xff] * int(self.width * self.height / 8))

        self.send_command(0x26)
        self.send_data2([0x00] * int(self.width * self.height / 8))
            
        self.TurnOnDisplay()
        
    # Turn on display
    def TurnOnDisplay(self):
        self.send_command(0x20)
        self.ReadBusy()

    # Enter sleep mode
    def sleep(self):
        self.send_command(0x10)
        self.send_data(0x01)
        
        epdconfig.delay_ms(2000)
        epdconfig.module_exit()
### END OF FILE ###

